### ISSUE TITLE
SA1002: SemicolonsMustBeSpacedCorrectly behaves differently from StyleCop's version regarding for loops

### ISSUE THREAD

**Issue by jackwhelpton** [DotNetAnalyzers/StyleCopAnalyzers#1489]
The implementation of this rule is a little different from the original StyleCop version when it comes to the sections of a "for" loop. In the previous version, you would be nudged towards writing an "infinite" for loop as:

````csharp
for (;;)
{
}
````

Other variants would trigger a violation of SA1002. By contrast, the only way to adhere to the rule now appears to be:

````csharp
for (; ;)
{
}
````

which raised a violation of SA1002 in the previous version. Is this deliberate? The same difference is apparent if you add just an initializer to the loop.

I'm posting several queries of this style as I migrate an existing solution to use the StyleCop Analyzers; is it generally intended that the rules that existed in the previous version should preserve their behaviour?

**Comment by sharwell** [comment]
The current form was chosen for its enhanced readability when only an initializer is given:

```csharp
for (int x = 3; ;)
{
}
```

However, even with the space the loop is not particularly readable in my mind. Eventually this led me to create #633.

:memo: Also note that I was not able to find many cases where this appeared in practice. For example, I found a total of two instances of this in the entire Roslyn code base (counting both `for (;;)` and `for (; ;)`).

> is it generally intended that the rules that existed in the previous version should preserve their behaviour?

Generally, yes. However, there are a number of intentional exceptions to this which we are trying to gather in #359 and eventually move to the documentation.

**Comment by jackwhelpton** [comment]
One nice thing about:

````csharp
for (int x = 3; ;)
{
}
````

over:

````csharp
int x = 3;
while (true)
{
}
````

is that the former construct allows you to limit the scope of the initialized variable. I suppose the exact analogue of this code as a ``while`` loop would be the rather heavily nested:

````csharp
{
    int x = 3;
    while (true)
    {
    }
}
````
